lib/charms/vault_k8s/v0/vault_client.py

```diff
@@ -416,3 +447,100 @@ def get_num_raft_peers(self) -> int:
         """Return the number of raft peers."""
         raft_config = self._client.sys.read_raft_config()
         return len(raft_config["data"]["config"]["servers"])
+
+    def _get_autounseal_policy_name(self, relation_id: int) -> str:
+        """Return the policy name for the given relation id."""
+        return f"charm-autounseal-{relation_id}"
+
+    def _get_autounseal_approle_name(self, relation_id: int) -> str:
+        """Return the approle name for the given relation id."""
+        return f"charm-autounseal-{relation_id}"
+
+    def _configure_autounseal_policy(self, mount: str, relation_id: int, key_name: str) -> str:
+        """Create/update a policy within vault to use the transit key."""
+        mount_policy = textwrap.dedent(f"""
+        path "{mount}/encrypt/{key_name}" {{
+            capabilities = ["update"]
+        }}
+
+        path "{mount}/decrypt/{key_name}" {{
+            capabilities = ["update"]
+        }}
+        """)
+        policy_name = self._get_autounseal_policy_name(relation_id)
+        self._client.sys.create_or_update_policy(
+            policy_name,
+            mount_policy,
+        )
+        return policy_name
+
+    def _get_autounseal_key_name(self, relation_id: int) -> str:
+        """Return the key name for the given relation id."""
+        return str(relation_id)
+
+    def _create_autounseal_key(self, mount_point: str, relation_id: int) -> str:
+        """Create a new autounseal key."""
+        key_name = self._get_autounseal_key_name(relation_id)
+        response = self._client.secrets.transit.create_key(mount_point=mount_point, name=key_name)
+        logging.debug(f"Created a new autounseal key: {response}")
+        return key_name
+
+    def _destroy_autounseal_key(self, mount_point, key_name):
+        """Destroy the autounseal key."""
+        self._client.secrets.transit.delete_key(mount_point=mount_point, name=key_name)
+
+    def destroy_autounseal_credentials(self, relation_id: int, mount: str) -> None:
```

Comment:
responded there but in short, lets keep it this way if we want to avoid cross lib dependencies.

Just as an aside, I didn't think cross lib dependencies would be an issue because at some point [Loki was doing it](https://github.com/canonical/loki-k8s-operator/commit/f8601e7a058ccf52c47cd86178aa8649b7f02893#diff-ad7051b5ca68befcdb77d638fcfa686fa902ae8089c8e72cea3c222df01a4597).

But it looks like they're moving away from that and publish a `cosl` package in pypi to share library dependencies now.

Reply:
yeah I had some discussion yesterday with @ghislainbourgeois and he mentioned it's possible, but the experience isn't great. Probably something we still want to discuss before we start.